src/meta/src/barrier/backfill_order_control.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use risingwave_pb::stream_plan::FragmentTypeFlag;
+
+use crate::model::{FragmentId, StreamJobFragments};
+
+/// This is the "global" `fragment_id`.
+/// The local `fragment_id` is namespaced by the `fragment_id`.
+pub type ActorId = u32;
+
+#[derive(Clone, Debug, Default)]
+pub struct BackfillNode {
+    fragment_id: FragmentId,
+    /// How many more actors need to finish,
+    /// before this fragment can finish backfilling.
+    remaining_actor_count: usize,
+    /// How many more dependencies need to finish,
+    /// before this fragment can be backfilled.
+    remaining_dependency_count: usize,
+    children: Vec<FragmentId>,
+}
+
+/// Actor done                   -> update `fragment_id` state
+/// Operator done                -> update downstream operator dependency
+/// Operator's dependencies done -> queue operator for backfill
+#[derive(Clone, Debug)]
+pub struct BackfillOrderState {
+    // The order plan.
+    current_backfill_nodes: HashMap<FragmentId, BackfillNode>,
+    // Remaining nodes to finish
+    remaining_backfill_nodes: HashMap<FragmentId, BackfillNode>,
+    // The mapping between actors and fragment_ids
+    actor_to_fragment_id: HashMap<ActorId, FragmentId>,
+}
+
+// constructor
+impl BackfillOrderState {
+    pub fn new(
+        backfill_orders: HashMap<FragmentId, Vec<FragmentId>>,
+        stream_job_fragments: &StreamJobFragments,
+    ) -> Self {
+        tracing::debug!(?backfill_orders, "initialize backfill order state");
+        let actor_to_fragment_id = stream_job_fragments.actor_fragment_mapping();
+
+        let mut backfill_nodes: HashMap<FragmentId, BackfillNode> = HashMap::new();
+
+        for fragment in stream_job_fragments.fragments() {
+            if fragment.fragment_type_mask
+                & (FragmentTypeFlag::StreamScan as u32 | FragmentTypeFlag::SourceScan as u32)
```

Comment:
Does backfill order control support source backfill? I don't see any changes to the `SourceBackfillExecutor`.

Reply:
Not in the initial parts, but eventually we should. So this PR only implements some easy parts for it.
Detailed in the Future Work section of the PR sescription